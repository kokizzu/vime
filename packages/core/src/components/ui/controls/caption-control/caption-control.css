vime-caption-control.hidden {
  display: none;
}