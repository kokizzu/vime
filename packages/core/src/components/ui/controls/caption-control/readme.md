# vime-caption-control

A control for toggling the visibility of captions. This control is not displayed if there's no track
currently set.

## Example

```html {6}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls>
      <vime-caption-control></vime-caption-control>
    </vime-controls>
  </vime-ui>
  <vime-player></vime-player
></vime-player>
```

<!-- Auto Generated Below -->

## Properties

| Property           | Attribute           | Description                                                                                                                                                   | Type                             | Default                |
| ------------------ | ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------- | ---------------------- |
| `hideIcon`         | `hide-icon`         | The URL to an SVG element or fragment to load.                                                                                                                | `string`                         | `'#vime-captions-off'` |
| `hideTooltip`      | `hide-tooltip`      | Whether the tooltip should not be displayed.                                                                                                                  | `boolean`                        | `false`                |
| `keys`             | `keys`              | A slash (`/`) seperated string of JS keyboard keys (`KeyboardEvent.key`), that when caught in a `keydown` event, will trigger a `click` event on the control. | `string \| undefined`            | `'c'`                  |
| `scale`            | `scale`             | Scale the size of the control up/down by the amount given.                                                                                                    | `number`                         | `1`                    |
| `showIcon`         | `show-icon`         | The URL to an SVG element or fragment to load.                                                                                                                | `string`                         | `'#vime-captions-on'`  |
| `tooltipDirection` | `tooltip-direction` | The direction in which the tooltip should grow.                                                                                                               | `"left" \| "right" \| undefined` | `undefined`            |

## Dependencies

### Used by

- [vime-default-controls](../default-controls)

### Depends on

- [vime-control](../control)
- [vime-icon](../../icon)
- [vime-tooltip](../../tooltip)

### Graph

```mermaid
graph TD;
  vime-caption-control --> vime-control
  vime-caption-control --> vime-icon
  vime-caption-control --> vime-tooltip
  vime-default-controls --> vime-caption-control
  style vime-caption-control fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
