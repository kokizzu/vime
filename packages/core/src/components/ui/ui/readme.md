# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Dependencies

### Used by

 - [vime-default-ui](../default-ui)

### Graph
```mermaid
graph TD;
  vime-default-ui --> vime-ui
  style vime-ui fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
