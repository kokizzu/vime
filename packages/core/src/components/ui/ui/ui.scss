vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}