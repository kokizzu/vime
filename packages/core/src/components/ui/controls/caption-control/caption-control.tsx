import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { PlayerProps, PlayerProp } from '../../../core/player/PlayerProp';
import { withPlayerContext } from '../../../core/player/PlayerContext';
import { TooltipDirection } from '../../tooltip/types';
import { isUndefined } from '../../../../utils/unit';
import { KeyboardControl } from '../control/KeyboardControl';
import { findRootPlayer } from '../../../core/player/utils';

@Component({
  tag: 'vime-caption-control',
  styleUrl: 'caption-control.css',
})
export class CaptionControl implements KeyboardControl {
  /**
   * The URL to an SVG element or fragment to load.
   */
  @Prop() showIcon = '#vime-captions-on';

  /**
   * The URL to an SVG element or fragment to load.
   */
  @Prop() hideIcon = '#vime-captions-off';

  /**
   * The direction in which the tooltip should grow.
   */
  @Prop() tooltipDirection: TooltipDirection;

  /**
   * Whether the tooltip should not be displayed.
   */
  @Prop() hideTooltip = false;

  /**
   * Scale the size of the control up/down by the amount given.
   */
  @Prop() scale = 1;

  /**
   * @inheritdoc
   */
  @Prop() keys?: string = 'c';

  /**
   * @internal
   */
  @Prop() currentCaption?: PlayerProps[PlayerProp.currentCaption];

  /**
   * @internal
   */
  @Prop() isCaptionsActive: PlayerProps[PlayerProp.isCaptionsActive] = false;

  /**
   * @internal
   */
  @Prop() i18n: PlayerProps[PlayerProp.i18n] = {};

  private onClick() {
    const player = findRootPlayer(this);
    player.toggleCaptionsVisiblity();
  }

  render() {
    const tooltip = this.isCaptionsActive ? this.i18n.disableCaptions : this.i18n.enableCaptions;
    const tooltipWithHint = !isUndefined(this.keys) ? `${tooltip} (${this.keys})` : tooltip;

    return (
      <Host
        class={{
          hidden: isUndefined(this.currentCaption),
        }}
      >
        <vime-control
          scale={this.scale}
          label={this.i18n.captions}
          keys={this.keys}
          hidden={isUndefined(this.currentCaption)}
          pressed={this.isCaptionsActive}
          onClick={this.onClick.bind(this)}
        >
          <vime-icon href={this.isCaptionsActive ? this.showIcon : this.hideIcon} />

          <vime-tooltip
            hidden={this.hideTooltip}
            direction={this.tooltipDirection}
          >
            {tooltipWithHint}
          </vime-tooltip>
        </vime-control>
      </Host>
    );
  }
}

withPlayerContext(CaptionControl, [
  PlayerProp.isCaptionsActive,
  PlayerProp.currentCaption,
  PlayerProp.i18n,
]);
