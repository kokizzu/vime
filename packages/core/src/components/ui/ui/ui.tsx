import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

/**
 * @slot - Used to pass in UI components for the player.
 */
@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.isVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.isFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        <div>
          {canShowCustomUI && <slot />}
        </div>
      </Host>
    );
  }
}

withPlayerContext(UI, [
  PlayerProp.isVideoView,
  PlayerProp.playsinline,
  PlayerProp.isFullscreenActive,
]);
