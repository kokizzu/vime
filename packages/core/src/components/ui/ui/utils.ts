import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
