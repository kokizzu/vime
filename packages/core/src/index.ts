import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
