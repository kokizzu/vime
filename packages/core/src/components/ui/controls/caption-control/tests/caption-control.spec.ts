import { SpecPage } from '@stencil/core/testing';
import { CaptionControl } from '../caption-control';
import { newUISpecPage } from '../../../ui/tests';

let page: SpecPage;
let control: HTMLVimeCaptionControlElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [CaptionControl],
    '<vime-caption-control />',
  ));

  control = page.root!.querySelector('vime-caption-control')!;
});

it('should be structurally sound when not active', () => {
  expect(control).toMatchSnapshot();
});

it('should be structurally sound when active', async () => {
  control.isCaptionsActive = true;
  await page.waitForChanges();
  expect(control).toMatchSnapshot();
});

it('should be hidden if there is no current caption', () => {
  expect(control.querySelector('vime-control')).toEqual({});
});

it('should not be hidden if there is a current caption', async () => {
  // @ts-ignore
  control.currentCaption = {};
  await page.waitForChanges();
  expect(control.querySelector('vime-control')?.nodeName).toEqual('VIME-CONTROL');
});
