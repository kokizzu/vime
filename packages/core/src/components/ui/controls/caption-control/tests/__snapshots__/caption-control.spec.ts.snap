// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound when active 1`] = `
<vime-caption-control class="hidden">
  <vime-control hidden="" keycodes="67" label="Captions" scale="1">
    <vime-icon href="#vime-captions-on"></vime-icon>
    <vime-tooltip>
      Disable subtitles/captions (c)
    </vime-tooltip>
  </vime-control>
</vime-caption-control>
`;

exports[`should be structurally sound when not active 1`] = `
<vime-caption-control class="hidden">
  <vime-control hidden="" keycodes="67" label="Captions" scale="1">
    <vime-icon href="#vime-captions-off"></vime-icon>
    <vime-tooltip>
      Enable subtitles/captions (c)
    </vime-tooltip>
  </vime-control>
</vime-caption-control>
`;
